### PR TITLE
fix: dx halo credential

### PR DESCRIPTION
**Pull request by egorgripasov**

### CODE REVIEW

File: packages/devtools/cli/src/base-command.ts [hunk 1]
```diff
@@ -80,6 +82,34 @@ export abstract class BaseCommand extends Command {
     this.log('ok');
   }
 
+  async getStdin() {
```
Comment (richburdon) [inline packages/devtools/cli/src/base-command.ts:85]: I thins https://nodejs.org/api/readline.html is the best way to read from stdin.
Move to util/readline.ts?

```
import readline from 'readline';

const rl = readline.createInterface({
  input: process.stdin,
  terminal: false
});

rl.on('line', (line) => {
    console.log(line);
});

rl.once('close', () => {
  // Done.
});
```

File: packages/devtools/cli/src/util/util.ts [hunk 2]
```diff
@@ -7,6 +7,7 @@ import { CliUx } from '@oclif/core';
 import { Space, SpaceMember } from '@dxos/client';
 import { truncateKey } from '@dxos/debug';
 import { PublicKey } from '@dxos/keys';
+import { Credential } from '@dxos/protocols/proto/dxos/halo/credentials';
 
 const maybeTruncateKey = (key: PublicKey, truncate = false) => (truncate ? truncateKey(key.toHex(), 8) : key.toHex());
```
Comment (richburdon) [inline packages/devtools/cli/src/util/util.ts:12]: truncateKey just takes a key

File: packages/devtools/cli/src/util/util.ts [hunk 3]
```diff
@@ -84,3 +85,35 @@ export const printMembers = (members: SpaceMember[], flags = {}) => {
     }
   );
 };
+
+export const mapCredentials = (credentials: Credential[], truncateKeys = false) => {
```
Comment (richburdon) [inline packages/devtools/cli/src/util/util.ts:89]: Refactor util file to co-locate all "print" related methods (always factor out "util" as it grows)

File: packages/devtools/cli/src/util/index.ts [hunk 4]
```diff
@@ -2,6 +2,7 @@
 // Copyright 2022 DXOS.org
 //
 
+export * from './credential';
```
Comment (richburdon) [inline packages/devtools/cli/src/util/index.ts:5]: sort imports

File: packages/devtools/cli/src/util/credential.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        result = credentials.filter(predicate);
+      } else {
+        result = credentials;
+      }
+    },
+    onError: (err) => {
+      throw err;
+    }
+  });
+  setTimeout(() => {
```
Comment (richburdon) [inline packages/devtools/cli/src/util/credential.ts:31]: Why is this required?

Comment (richburdon) [reply packages/devtools/cli/src/util/credential.ts:31]: @wittjosiah should this be halo.credentials.get?

**Review by richburdon — CHANGES_REQUESTED** [review]

File: packages/devtools/cli/src/util/credential.ts (hunk 5 above)
Comment (egorgripasov) [reply packages/devtools/cli/src/util/credential.ts:31]: this is required because we are reading creds from feed, and they are coming progressively (not all at once).
e.g. if you have 5 creds in your halo, you will get 5 `onUpdate` events:
1 - with 1 cred
2 - with 2 creds
...
5 - with all 5 creds.
Since we don't really know how many credentials will be there, we could only wait for some time till we think all of them are returned

File: packages/devtools/cli/src/util/util.ts (hunk 3 above)
Comment (egorgripasov) [reply packages/devtools/cli/src/util/util.ts]: +1

File: packages/devtools/cli/src/util/credential.ts (hunk 5 above)
Comment (mykola-vrmchk) [reply packages/devtools/cli/src/util/credential.ts:31]: We are working on this with @dmaretskyi, so for now timeout is ok
